runtime/vcache/primitive.go

```diff
@@ -1,68 +1,144 @@
+	// all at once, compute the byte offsets of each value
+	// (for random access, not used yet).
+	var n int
+	for _, segment := range m.Segmap {
+		n += int(segment.MemLength)
+	}
+	bytes := make([]byte, n)
+	var off int
+	for _, segment := range m.Segmap {
+		if err := segment.Read(r, bytes[off:]); err != nil {
+			return nil, err
 		}
-		data := make([]byte, n)
-		var off int
-		for _, segment := range p.meta.Segmap {
-			if err := segment.Read(r, data[off:]); err != nil {
-				return nil, err
+		off += int(segment.MemLength)
+	}
+	switch typ := typ.(type) {
+	case *zed.TypeOfUint8, *zed.TypeOfUint16, *zed.TypeOfUint32, *zed.TypeOfUint64, *zed.TypeOfTime:
+		//XXX put valcnt in vng meta and use vector allocator
+		var vals []uint64
+		var nullslots []uint32
+		it := zcode.Bytes(bytes).Iter()
+		for !it.Done() {
+			val := it.Next()
+			if val == nil {
+				nullslots = append(nullslots, uint32(len(vals)))
+				vals = append(vals, 0)
+			} else {
+				vals = append(vals, zed.DecodeUint(val))
 			}
-			off += int(segment.MemLength)
 		}
-		p.bytes = data
-	}
-	if dict := p.meta.Dict; dict != nil {
-		bytes := p.bytes
-		return func(b *zcode.Builder) error {
-			pos := bytes[0]
-			bytes = bytes[1:]
-			b.Append(dict[pos].Value.Bytes())
-			return nil
-		}, nil
+		return vector.NewUint(typ, vals, vector.NewNullmask(nullslots, len(vals))), nil
+	case *zed.TypeOfInt8, *zed.TypeOfInt16, *zed.TypeOfInt32, *zed.TypeOfInt64, *zed.TypeOfDuration:
+		//XXX put valcnt in vng meta and use vector allocator
+		var vals []int64
+		var nullslots []uint32
+		it := zcode.Bytes(bytes).Iter()
+		for !it.Done() {
+			val := it.Next()
+			if val == nil {
+				nullslots = append(nullslots, uint32(len(vals)))
+				vals = append(vals, 0)
+			} else {
+				vals = append(vals, zed.DecodeInt(val))
+			}
+		}
+		return vector.NewInt(typ, vals, vector.NewNullmask(nullslots, len(vals))), nil
+	case *zed.TypeOfFloat16:
+		return nil, fmt.Errorf("vcache.Primitive.Load TBD for %T", typ)
+	case *zed.TypeOfFloat32:
+		return nil, fmt.Errorf("vcache.Primitive.Load TBD for %T", typ)
+	case *zed.TypeOfFloat64:
+		return nil, fmt.Errorf("vcache.Primitive.Load TBD for %T", typ)
+	case *zed.TypeOfBool:
+		var vals []bool
+		var nullslots []uint32
+		it := zcode.Bytes(bytes).Iter()
+		for !it.Done() {
+			val := it.Next()
+			if val == nil {
+				nullslots = append(nullslots, uint32(len(vals)))
+				vals = append(vals, false)
+			} else {
+				vals = append(vals, zed.DecodeBool(val))
+			}
+		}
+		return vector.NewBool(typ, vals, vector.NewNullmask(nullslots, len(vals))), nil
+	case *zed.TypeOfBytes:
+		return nil, fmt.Errorf("vcache.Primitive.Load TBD for %T", typ)
+	case *zed.TypeOfString:
+		var vals []string
+		var nullslots []uint32
+		it := zcode.Bytes(bytes).Iter()
+		for !it.Done() {
+			val := it.Next()
+			if val == nil {
+				nullslots = append(nullslots, uint32(len(vals)))
+			} else {
+				vals = append(vals, zed.DecodeString(val))
```

Comment:
`zed.DecodeString(val)`

If I'm reading this right this is not a copy - all the resulting strings share the same underlying byte buffer?

Reply:
That's what we want but there's an implicit copy here as the compiler will insert a call to runtime.slicebytetostring when it sees this. Our usual way around that is this: https://github.com/brimdata/zed/blob/40639dce6c333da1c55844b026052dc3718cc2d2/pkg/byteconv/byteconv.go#L11-L16